kernel/src/mm/guestmem.rs

```diff
@@ -269,6 +278,98 @@ impl<T: Copy> InsnMachineMem for GuestPtr<T> {
     }
 }
 
+struct UserAccessGuard;
+
+impl Drop for UserAccessGuard {
+    fn drop(&mut self) {
+        clac();
+    }
+}
+
+fn user_access_begin() -> UserAccessGuard {
+    stac();
+    UserAccessGuard
+}
+
+#[derive(Debug)]
+pub struct UserPtr<T: Copy> {
+    guest_ptr: GuestPtr<T>,
+}
+
+impl<T: Copy> UserPtr<T> {
+    #[inline]
+    pub fn new(v: VirtAddr) -> Self {
+        Self {
+            guest_ptr: GuestPtr::new(v),
+        }
+    }
+
+    fn check_bounds(&self) -> bool {
+        let v = VirtAddr::from(self.guest_ptr.ptr);
+
+        (USER_MEM_START..USER_MEM_END).contains(&v)
+            && (USER_MEM_START..USER_MEM_END).contains(&(v + size_of::<T>()))
+    }
+
+    #[inline]
+    pub fn read(&self) -> Result<T, SvsmError>
+    where
+        T: FromBytes,
+    {
+        let _guard = user_access_begin();
+        if !self.check_bounds() {
+            return Err(SvsmError::InvalidAddress);
+        }
```

Comment:
```suggestion
        if !self.check_bounds() {
            return Err(SvsmError::InvalidAddress);
        }
        let _guard = user_access_begin();
```

The same change should be applied for `write_ref`.

Reply:
Since we switched to `new` method, used it here rather than calling the function.